File: android/src/main/java/com/github/quadflask/react/navermap/RNNaverMapPathOverlayManager.java
package com.github.quadflask.react.navermap;

import android.content.Context;
import android.graphics.Color;
import android.os.Build;
import android.util.DisplayMetrics;
import android.view.WindowManager;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.uimanager.ThemedReactContext;
import com.facebook.react.uimanager.ViewGroupManager;
import com.facebook.react.uimanager.annotations.ReactProp;

import javax.annotation.Nonnull;

import static com.github.quadflask.react.navermap.ReactUtil.toLatLngList;

public class RNNaverMapPathOverlayManager extends ViewGroupManager<RNNaverMapPathOverlay> {
    private final DisplayMetrics metrics;

    public RNNaverMapPathOverlayManager(ReactApplicationContext reactContext) {
        super();
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
            metrics = new DisplayMetrics();
            ((WindowManager) reactContext.getSystemService(Context.WINDOW_SERVICE))
                    .getDefaultDisplay()
                    .getRealMetrics(metrics);
        } else {
            metrics = reactContext.getResources().getDisplayMetrics();
        }
    }

    @Nonnull
    @Override
    public String getName() {
        return "RNNaverMapPathOverlay";
    }

    @Nonnull
    @Override
    protected RNNaverMapPathOverlay createViewInstance(@Nonnull ThemedReactContext reactContext) {
        return new RNNaverMapPathOverlay(reactContext);
    }

    @ReactProp(name = "coordinates")
    public void setCoordinate(RNNaverMapPathOverlay view, ReadableArray coordinates) {
        view.setCoords(toLatLngList(coordinates));
    }

    @ReactProp(name = "width", defaultFloat = 1f)
    public void setStrokeWidth(RNNaverMapPathOverlay view, float widthInPoints) {
        float widthInScreenPx = metrics.density * widthInPoints;
        view.setWidth(widthInScreenPx);
    }

    @ReactProp(name = "color", defaultInt = Color.RED, customType = "Color")
    public void setColor(RNNaverMapPathOverlay view, int color) {
        view.setColor(color);
    }

    @ReactProp(name = "outlineWidth", defaultFloat = 1f)
    public void setOutlineWidth(RNNaverMapPathOverlay view, float widthInPoints) {
        float widthInScreenPx = metrics.density * widthInPoints;
        view.setOutlineWidth(widthInScreenPx);
    }

    @ReactProp(name = "passedColor", defaultInt = Color.RED, customType = "Color")
    public void setPassedColor(RNNaverMapPathOverlay view, int color) {
        view.setPassedColor(color);
    }

    @ReactProp(name = "outlineColor", defaultInt = Color.RED, customType = "Color")
    public void setOutlineColor(RNNaverMapPathOverlay view, int color) {
        view.setOutlineColor(color);
    }

    @ReactProp(name = "passedOutlineColor", defaultInt = Color.RED, customType = "Color")
    public void setPassedOutlineColor(RNNaverMapPathOverlay view, int color) {
        view.setPassedOutlineColor(color);
    }

    @ReactProp(name = "pattern")
    public void setPatternInterval(RNNaverMapPathOverlay view, String image) {
        view.setPattern(image);
    }

    @ReactProp(name = "patternInterval", defaultFloat = 1f)
    public void setPatternInterval(RNNaverMapPathOverlay view, float widthInPoints) {
        int widthInScreenPx = Math.round(metrics.density * widthInPoints);
        view.setPatternInterval(widthInScreenPx);
    }
}

File: android/src/main/java/com/github/quadflask/react/navermap/RNNaverMapPathOverlayManager.kt
package com.github.quadflask.react.navermap

import android.graphics.Color
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReadableArray
import com.facebook.react.uimanager.ThemedReactContext
import com.facebook.react.uimanager.annotations.ReactProp
import com.github.quadflask.react.navermap.ReactUtil.toLatLngList
import kotlin.math.roundToInt

class RNNaverMapPathOverlayManager(reactContext: ReactApplicationContext) : EventEmittableViewGroupManager<RNNaverMapPathOverlay>(reactContext) {
  override val eventNames = arrayOf("onClick")

  override fun getName() = "RNNaverMapPathOverlay"

  override fun createViewInstance(reactContext: ThemedReactContext): RNNaverMapPathOverlay {
    return RNNaverMapPathOverlay(this, reactContext)
  }

  @ReactProp(name = "coordinates")
  fun setCoordinate(view: RNNaverMapPathOverlay, coordinates: ReadableArray?) {
    view.setCoords(toLatLngList(coordinates))
  }

  @ReactProp(name = "width", defaultFloat = 1f)
  fun setStrokeWidth(view: RNNaverMapPathOverlay, widthInPoints: Float) {
    val widthInScreenPx = metrics.density * widthInPoints
    view.setWidth(widthInScreenPx)
  }

  @ReactProp(name = "zIndex", defaultInt = 0)
  fun setProgress(view: RNNaverMapPathOverlay, zIndex: Int) {
    view.setZIndex(zIndex)
  }

  @ReactProp(name = "color", defaultInt = Color.RED, customType = "Color")
  fun setColor(view: RNNaverMapPathOverlay, color: Int) {
    view.setColor(color)
  }

  @ReactProp(name = "outlineWidth", defaultFloat = 1f)
  fun setOutlineWidth(view: RNNaverMapPathOverlay, widthInPoints: Float) {
    val widthInScreenPx = metrics.density * widthInPoints
    view.setOutlineWidth(widthInScreenPx)
  }

  @ReactProp(name = "passedColor", defaultInt = Color.RED, customType = "Color")
  fun setPassedColor(view: RNNaverMapPathOverlay, color: Int) {
    view.setPassedColor(color)
  }

  @ReactProp(name = "outlineColor", defaultInt = Color.RED, customType = "Color")
  fun setOutlineColor(view: RNNaverMapPathOverlay, color: Int) {
    view.setOutlineColor(color)
  }

  @ReactProp(name = "passedOutlineColor", defaultInt = Color.RED, customType = "Color")
  fun setPassedOutlineColor(view: RNNaverMapPathOverlay, color: Int) {
    view.setPassedOutlineColor(color)
  }

  @ReactProp(name = "pattern")
  fun setPatternInterval(view: RNNaverMapPathOverlay, image: String?) {
    view.setPattern(image)
  }

  @ReactProp(name = "patternInterval", defaultFloat = 1f)
  fun setPatternInterval(view: RNNaverMapPathOverlay, widthInPoints: Float) {
    val widthInScreenPx = (metrics.density * widthInPoints).roundToInt()
    view.setPatternInterval(widthInScreenPx)
  }

  @ReactProp(name = "progress", defaultFloat = 0f)
  fun setProgress(view: RNNaverMapPathOverlay, progress: Float) {
    view.setProgress(progress)
  }
}
